### PR TITLE
#2801 Fixed unit test for CSRFCountermeasures

### PR DESCRIPTION
**Pull request by peterschuler**
Added unit test for CSRFCountermeasures, Minor changes to CSRFCountermeasures to facilitate testing.

Added the unit test. 

Fixes zaproxy/zaproxy#2801

### CODE REVIEW

File: src/org/zaproxy/zap/extension/pscanrulesBeta/CSRFCountermeasures.java [hunk 1]
```diff
@@ -104,15 +104,14 @@ public void scanHttpResponseReceive(HttpMessage msg, int id, Source source) {
...
-			(ExtensionAntiCSRF) Control.getSingleton().getExtensionLoader().getExtension(ExtensionAntiCSRF.NAME);
+		ExtensionAntiCSRF extAntiCSRF = getExtensonAntiCSRF();
 		
 		if (extAntiCSRF == null) {
 			return;
 		}
 		
-		List<Element> formElements = source.getAllElements(HTMLElementName.FORM);
 		List<String> tokenNames = extAntiCSRF.getAntiCsrfTokenNames();
+		List<Element> formElements = source.getAllElements(HTMLElementName.FORM);
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/pscanrulesBeta/CSRFCountermeasures.java]: Why is this change needed?

File: src/org/zaproxy/zap/extension/pscanrulesBeta/CSRFCountermeasures.java [hunk 2]
```diff
@@ -104,15 +104,14 @@ public void scanHttpResponseReceive(HttpMessage msg, int id, Source source) {
 		
 		long start = System.currentTimeMillis();
 		
-		ExtensionAntiCSRF extAntiCSRF = 
-			(ExtensionAntiCSRF) Control.getSingleton().getExtensionLoader().getExtension(ExtensionAntiCSRF.NAME);
+		ExtensionAntiCSRF extAntiCSRF = getExtensonAntiCSRF();
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/pscanrulesBeta/CSRFCountermeasures.java]: Extension is misspelt

File: src/org/zaproxy/zap/extension/pscanrulesBeta/CSRFCountermeasures.java (hunk 1 above)
Comment (peterschuler) [inline src/org/zaproxy/zap/extension/pscanrulesBeta/CSRFCountermeasures.java]: Not needed. I think the lines where unintentionally swapped. Only the order changed. Want me to revert this?

File: src/org/zaproxy/zap/extension/pscanrulesBeta/CSRFCountermeasures.java (hunk 2 above)
Comment (peterschuler) [inline src/org/zaproxy/zap/extension/pscanrulesBeta/CSRFCountermeasures.java]: fixed.

File: src/org/zaproxy/zap/extension/pscanrulesBeta/CSRFCountermeasures.java (hunk 1 above)
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/pscanrulesBeta/CSRFCountermeasures.java]: Yes please

Comment (peterschuler) [inline src/org/zaproxy/zap/extension/pscanrulesBeta/CSRFCountermeasures.java]: done.

File: test/org/zaproxy/zap/extension/pscanrulesBeta/CSRFCountermeasuresTest.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
+package org.zaproxy.zap.extension.pscanrulesBeta;
```
Comment (kingthorin) [inline test/org/zaproxy/zap/extension/pscanrulesBeta/CSRFCountermeasuresTest.java]: Should have the standard license

File: test/org/zaproxy/zap/extension/pscanrulesBeta/TestableCSRFCountermeasures.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
+package org.zaproxy.zap.extension.pscanrulesBeta;
```
Comment (kingthorin) [inline test/org/zaproxy/zap/extension/pscanrulesBeta/TestableCSRFCountermeasures.java]: Should have the standard license

**Comment by psiinon** [comment]
Hi @peterschuler - this looks good to me. Can you add our std licence headers as requested by @kingthorin ? Or "Allow edits from maintainers" so we can do it for you - its a minor change so you'll still qualify for the bounty :)

**Comment by peterschuler** [comment]
Nah, I will do it this weekend.

Simon Bennetts <notifications@github.com>schreef:

**Comment by kingthorin** [comment]
Any news on this?

**Comment by kingthorin** [comment]
@peterschuler Please do an interactive rebase and fix-up or squash into a single "atomic" commit: https://github.com/zaproxy/zaproxy/blob/develop/CONTRIBUTING.md#what-we-zap-team-expect-from-you .

- https://www.kernel.org/pub/software/scm/git/docs/git-rebase.html#_interactive_mode
- https://wiki.eclipse.org/EGit/User_Guide#Interactive_Rebase

**Comment by kingthorin** [comment]
TODO:

- [ ] `getExtensionAntiCSRF()` should use class method instead of name method, avoiding the cast.
- [ ] Fixup/squash
- [ ] TestableCSRFCountermeasures could be removed and the scanner just changed to allow to set the "ignoreConf" String (allow to set instead of extend, if the String was set use that otherwise use `Model.getSingleton().getOptionsParam().getConfig().getString("rules.csrf.ignorelist")`)
  - [ ] same for the extension
- [ ] does.not.matter.com should be example.com and use the constructor URI(String, boolean)

**Comment by kingthorin** [comment]
TODOs finish, unfortunately now it's failing. During the run of the test it attempts to create a working directory and fails.

> Failed to create directory C:\Users\user\OWASP ZAP_D\C:\Users\user\OWASP ZAP_D\plugin
> Failed to create directory C:\Users\user\OWASP ZAP_D\C:\Users\user\OWASP ZAP_D\C:\Users\user\OWASP ZAP_D\plugin

Will look into it further tonight.

**Comment by kingthorin** [comment]
Seems to have something to do with resource messages. So either I made a mistake when resolving conflicts in the scanner code itself, or something is lacking in our mocking/setup for the tests.

More later.

**Comment by thc202** [comment]
That's caused by the (new) methods `getCSRFIgnoreAttName()` and `getCSRFIgnoreAttValue()`, they need to be changed to be like `getCSRFIgnoreList()` (guess we also could allow to set a `Configuration` and all 3 methods changed to use that if non-null?).

**Comment by kingthorin** [comment]
Issue resolved. **Note**: I also:

- Tweaked all the method names. They used to be similar to `shouldPassWhenThereIsOnlyOneFormWithSecondKnownCSRFTokenUsingName` obviously the tests should pass. That one became `shouldNotRaiseAlertWhenThereIsOnlyOneFormWithSecondKnownCSRFTokenUsingName`.
- Added a few more tests to address the new security annotations rule config behavior.

File: src/org/zaproxy/zap/extension/pscanrulesBeta/CSRFCountermeasures.java [hunk 5]
```diff
@@ -63,6 +63,10 @@
      * the passive scan thread class used
      */
 	private PassiveScanThread parent = null;
+	private ExtensionAntiCSRF extensionAntiCSRF;
+	private static String csrfIgnoreList;
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/pscanrulesBeta/CSRFCountermeasures.java]: This (and following) should be instance variables.

File: test/org/zaproxy/zap/extension/pscanrulesBeta/CSRFCountermeasuresUnitTest.java [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+		HttpRequestHeader requestHeader = new HttpRequestHeader();
+		requestHeader.setURI(new URI("http://example.com", false));
+		
+		msg = new HttpMessage();
+		msg.setRequestHeader(requestHeader);
+	}
+	
+	@After
+	public void after() {
+		((CSRFCountermeasures)rule).setCsrfIgnoreList("");
```
Comment (thc202) [inline test/org/zaproxy/zap/extension/pscanrulesBeta/CSRFCountermeasuresUnitTest.java]: This (and following) should be done in the method `before()`.

**Comment by thc202** [comment]
@peterschuler, @kingthorin thanks!

**Comment by kingthorin** [comment]
@psiinon, leaving to you for merge since I ended up working on it.
